lib/googleauth/bearer_token.rb

```diff
@@ -0,0 +1,157 @@
+      # @option options [String] :token The bearer token to use.
+      # @option options [Time, Numeric, nil] :expiry The token expiry time provided by the end-user.
+      #   Optional, for the end-user's convenience. Can be a Time object, a number of seconds since epoch.
+      #   If the expiry is `nil`, it is treated as "token never expires".
+      # @option options [Symbol] :token_type The token type. Allowed values are
+      #   :access_token, :jwt, :id_token, and :bearer_token. Defaults to :bearer_token.
+      # @option options [String] :universe_domain The universe domain of the universe
+      #   this token is for (defaults to googleapis.com)
+      def initialize options = {}
+        raise ArgumentError, "Bearer token must be provided" if options[:token].nil? || options[:token].empty?
+        @token = options[:token]
+        @expiry = if options[:expiry].is_a? Time
+                    options[:expiry]
+                  elsif options[:expiry].is_a? Numeric
+                    Time.at options[:expiry]
+                  end
+
+        @token_type = options[:token_type] || :bearer_token
+        unless ALLOWED_TOKEN_TYPES.include? @token_type
+          raise ArgumentError, "Invalid token type: #{@token_type}. Allowed values are #{ALLOWED_TOKEN_TYPES.inspect}"
+        end
+
+        @universe_domain = options[:universe_domain] || "googleapis.com"
+      end
+
+      # Determines if the credentials object has expired.
+      #
+      # @param [Numeric] seconds The optional timeout in seconds.
+      # @return [Boolean] True if the token has expired, false otherwise, or
+      #   if the expiry was not provided.
+      def expires_within? seconds
+        return false if @expiry.nil? # Treat nil expiry as "never expires"
+        Time.now + seconds >= @expiry
+      end
+
+      # Creates a duplicate of these credentials.
+      #
+      # @param [Hash] options Additional options for configuring the credentials
+      # @option options [String] :token The bearer token to use.
+      # @option options [Time, Numeric] :expiry The token expiry time. Can be a Time
+      #   object or a number of seconds since epoch.
+      # @option options [Symbol] :token_type The token type. Allowed values are
+      #   :access_token, :jwt, :id_token, and :bearer_token. Defaults to :bearer_token.
+      # @option options [String] :universe_domain The universe domain (defaults to googleapis.com)
+      # @return [Google::Auth::BearerTokenCredentials]
+      def duplicate options = {}
+        self.class.new(
+          token: options[:token] || @token,
+          expiry: options[:expiry] || @expiry,
+          token_type: options[:token_type] || @token_type,
+          universe_domain: options[:universe_domain] || @universe_domain
+        )
+      end
+
+      protected
+
+      # We don't need to fetch access tokens for bearer token auth
+      def fetch_access_token! _options = {}
+        @token
```

Comment:
Here's a design question. If the user sets an expiration time, clients might automatically detect near-expiration (using `expires_within?`) and then call `fetch_access_token` with the understanding that it will refresh the token. This method, however, doesn't do so, and "fails silently" about it. Would it make sense to do something more serious like raise an error? Or at least log a warning? Or maybe do so only if fetch is called more than once (in case some clients proactively call fetch before a credential is first used)? I imagine that some clients might want to subclass this class and implement their own token refresh mechanism by overriding `fetch_access_token!`, and this would also help signal that the original nonfunctional fetch is still in operation.

Reply:
Rewrote `fetch_access_token!` to throw on expiry